build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1312,6 +1323,57 @@ message OutputSymlink {
   NodeProperties node_properties = 4;
 }
 
+// The termination result of a process, as reported by the operating
+// system to the parent process (e.g., using wait*() on POSIX-like
+// systems, or GetExitCodeProcess() on Microsoft Windows).
+//
+// Consumers SHOULD NOT assume that `reason` is exhaustive. Additional
+// operating system specific methods of process termination may be
+// covered by future versions of this protocol. Consumers MAY assume
+// that any future method indicates unsuccessful process termination.
+message ProcessTerminationResult {
+  message Exited {
+    // The exit code of the process.
+    //
+    // Implementations SHOULD preserve as many bits from the exit code
+    // as possible. On POSIX-like systems, it is recommended to use
+    // waitid(), as siginfo_t's si_status field contains the full exit
+    // code. Functions such as wait() and waitpid() tend to truncate the
+    // exit code to the bottom eight bits.
+    int64 exit_code = 1;
+  }
+
+  message Signaled {
+    // The name of the signal that was delivered that caused the process
+    // to terminate, without the "SIG" prefix (e.g., "BUS", "KILL",
+    // "SEGV").
+    string signal_name = 1;
+  }
+
+  oneof reason {
```

Comment:
How would you feel about not using a oneof and simply having a set of optional fields? signal_name implying the lack of exit_code makes sense in this particular case, but I'm not sure every future result is going to follow the pattern of being mutually exclusive. E.g. if we also provided info from the syslog or Windows Event Log so you could know _why_ the OS killed the process, say, where you'd get both a SIGKILL and some indication that it was killed due to OOM or too many handles or whatever.

So e.g.
```
message ProcessTerminationResult {
    //...
    int64 exit_code = 1;
    //...
    string signal_name = 2;
    //...
    string human_readable_explanation_for_signal = 3; 
}
```

With comments detailing the relationship between the various fields.

Reply:
Would that open the door to problems with default values? I'm thinking about `exit_code` in particular.

Reply:
Hmm, good point. For many the empty field is trivially uninteresting (e.g. string signal, empty==missing) but you're right about exit_code.

A few options:
 1. Do nothing, since exit_code can be distinguished fairly readily: the only case of concern is when it's 0, where we shouldn't actually need to distinguish if its set or not. (Do we really need to provide a ProcessTerminationResult just to provide a 0 exit_code? Given that the top-level field already exists, it seems just as reasonable to omit the whole message entirely in that case. So "0 means unset" for exit_code is just fine, since it introduces no real ambiguity in practice.)
 1. Do something special for exit_code, since so far it's the only one that has an ambiguous default. A second "has_exit_code" field would do, for example.
 1. If you know a set that will definitely be mutually exclusive, a oneof suffices, and then we can use optional fields for the rest. So e.g. exit_code and signal_name within the oneof, but with the intent for augmenting fields (like the human_readable_explanation_for_signal example) to be optionals outside of it. Which I guess is...what you already have, just with the understanding that maybe not everything will go in the oneof?
 1. Make these structs with a single field instead of raw fields, since you _can_ check for the presence of a struct. E.g.

```
message ProcessTerminationResult {
    message ExitCode {
        //...
        int64 code = 1;
    }
    ExitCode exit_code = 1;
    message Signal {
        //...
        string name = 1;
        string human_readable_explanation_for_signal = 2; 
    }
    Signal signal = 2;
}
```

I think I vote for (1): doesn't really feel like there's "real" ambiguity being introduced in this case, even for exit_code.